### PR TITLE
kubectl debug: add sysadmin profile

### PR DESCRIPTION
**Pull request by eiffel-fl**
Add the sysadmin profile from KEP 1441
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#debugging-profiles

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds the `sysadmin` profile as designed in [KEP 1441](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#debugging-profiles).

#### Which issue(s) this PR fixes:

https://github.com/kubernetes/kubectl/issues/1108

#### Does this PR introduce a user-facing change?

```release-note
kubectl debug: add sysadmin profile
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#debugging-profiles
```

### CODE REVIEW

**Comment by alban** [comment]
Thanks @eiffel-fl.

If you have your branch compiled, could you print the effective caps given by the netadmin and sysadmin profiles?
```
$ kubectl debug --profile=netadmin node/minikube-docker -ti --image=busybox -- grep CapEff /proc/self/status
CapEff:	0000003fffffffff
$ kubectl debug --profile=sysadmin node/minikube-docker -ti --image=busybox -- grep CapEff /proc/self/status
?
```
I'd like to make sure the privileges are really given in practice.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:194]: Could you sort these?

**Review by verb — APPROVED** [review]
Thanks for adding the sysadmin profile!

/lgtm

**Comment by verb** [comment]
@mochizuki875 Do you foresee any issues for CRIO similar to #118962?

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles_test.go
Comment (verb) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles_test.go:894]: whoops, I think this should be `sysadminProfile`

/remove-lgtm

**Comment by mochizuki875** [comment]
@verb 
To be honest, I'm not sure what the sysadmin profile is used for and expected to compared to netadmin...
However, at least CRI-O seems to have 5 fewer Capabilities by default than containerd.
So containers run by CRI-O should have fewer privileges compared to Containerd.

The difference is as below:
```
CAP_MKNOD
CAP_NET_RAW
CAP_SETPCAP
CAP_SYS_CHROOT
CAP_AUDIT_WRITE
```

Containerd's default Capabilities
https://github.com/containerd/containerd/blob/70a2c95ae8c02d7a4e448f0e4fb8bb0e6344b5c7/oci/spec.go#L118-L135

CRI-O's default Capabilities
https://github.com/cri-o/cri-o/blob/e81d41119d602a4c7e42a28a3c2bee42accae671/install.md?plain=1#L606

And as mentiond [here](https://github.com/kubernetes/kubernetes/pull/119200#issuecomment-1632539403), the sysadmin profile in case of node adds privileged [just like netadmin](https://github.com/kubernetes/kubernetes/issues/118962#issuecomment-1624777533).
So I think we need to decide whether to add all Capabilities by adding Privileged or just sysadmin.

**Comment by mochizuki875** [comment]
@eiffel-fl 
I don't fully understand `ig trace exec`, is it something that only depends on `SYS_ADMIN` regardless op `privileged: true`?
`privileged: true` not only adds all Capabilities, but also give various privileges to the container process.
(ex bypass security check like seccomp, allow to access some dir like `/proc/sys`) 

IMO: I think this fix should add integration test case like other profile. What do you think?
https://github.com/kubernetes/kubernetes/blob/master/test/cmd/debug.sh

**Comment by mochizuki875** [comment]
@eiffel-fl 
When you add new tests, you need to add to `legacy-script.sh` as described [here](https://github.com/kubernetes/kubernetes/tree/master/test/cmd#adding-tests).

**Comment by mochizuki875** [comment]
> I think I cannot add the sysadmin integration test for node as the cluster was not configured to allow privileged containers.

It comes from no `--allow-privileged=true` when starting kube-apiserver.
It may work with this flag.
However, I think we should decide whether sysadmin profile need privileged at first as mentioned [here](https://github.com/kubernetes/kubernetes/pull/119200#issuecomment-1693700790).

https://github.com/kubernetes/kubernetes/blob/7e9fbc449ddccab5be18d7d5fa0a4158ec8227f2/hack/make-rules/test-cmd.sh#L67-L84

**Comment by verb** [comment]
> @verb To be honest, I'm not sure what the sysadmin profile is used for and expected to compared to the case of netadmin... However, at least CRI-O seems to have 5 fewer Capabilities by default than containerd. So containers run by CRI-O should have fewer privileges compared to Containerd.

The intention of `sysadmin` was to node-level debugging to cluster administrators for the cases where the node is perhaps running a stripped-down OS that's lacking debugging tools. It's been a while since I've had to ssh to a node, though, so I don't remember the common operations. In the least it should include inspecting files, viewing processes, and entering other namespaces.

Maybe it would be better to start with a smaller set of capabilities and expand them as use cases arise, though.

**Comment by mochizuki875** [comment]
If the only capability we really need is SYS_ADMIN, I don't think we need to set privileged here.
Also, if we need "privileged" for the "sysadmin" profile, I think the profile name "sysadmin" is not appropriate and we should rename it or create a new profile like "privileged" profile.

**Comment by lion7** [comment]
I'm watching this issue specifically because I want to use the `sysadmin` profile to debug my Talos nodes.
Talos Linux is a kubernetes-specific OS that has no shell, sshd or any other traditional tools which should make it more secure. But it also makes debugging node-related issues very hard. 

Currently I mostly use the `netadmin` profile for my debugging, since it already applies privileged (I wonder why though). But I do still have to mount the root filesystem every time I start a new debugging pod, which is more of an annoyance than a real issue. Anyway, I'd love to have a more dedicated profile for debugging the node itself.

Side question: I see removing the privileged flag is considered. Would that prevent doing stuff like mounting raw devices / running `dmidecode` / other hardware related operations? Or is that already allowed by the `SYS_ADMIN` capability? 
In my case I know that Talos uses the `RuntimeDefault` seccomp profile which prevents some stuff like creating user namespaces; if removing the privileged flag would still enforce seccomp it would be a bit of a bummer. A dedicated `privileged` profile would obviously fit that use-case better, but I'm not sure how viable that proposal is.

**Comment by mochizuki875** [comment]
@lion7 
I think this use case makes sense for `kubectl debug`.
However, using `netadmin` profile for this case seems to not appropriate because `netadmin` profile is for network debugging(ex `tcpdump`, `mtr`) not debugging as administrator.
Currently, `privileged` is applied to `netadmin` profile, but it is not appropriate as you say and considered to be removed.

> Would that prevent doing stuff like mounting raw devices / running dmidecode / other hardware related operations? Or is that already allowed by the SYS_ADMIN capability?

I can't say for sure, but I think `mount` operation is within the privileges of `SYS_ADMIN` at least.
And I think `RuntimeDefault` seccomp profile will not applied by default unless you have set it in `securityContext` or enabled the [SeccompDefault feature](https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads). 

You can try it out by deploying a pod corresponding to `netadmin` or `sysadmin` profile for debugging node with these capabilities and see if it works.

`netadmin.yaml`
<details>

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: netadmin
spec:
  hostIPC: true
  hostNetwork: true
  hostPID: true
  containers:
  - image: <YOUR IMAGE>
    name: debugger
    securityContext:
      capabilities:
        add:
        - NET_ADMIN
        - NET_RAW
    volumeMounts:
    - mountPath: /host
      name: host-root
  volumes:
  - hostPath:
      path: /
      type: ""
    name: host-root
```

</details>

`sysadmin.yaml`(removed `privileged`)
<details>

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: sysadmin
spec:
  hostIPC: true
  hostNetwork: true
  hostPID: true
  containers:
  - image: <YOUR IMAGE>
    name: debugger
    securityContext:
      capabilities:
        add:
        - SYS_ADMIN
    volumeMounts:
    - mountPath: /host
      name: host-root
  volumes:
  - hostPath:
      path: /
      type: ""
    name: host-root
```

</details>


> A dedicated privileged profile would obviously fit that use-case better, but I'm not sure how viable that proposal is.

I agree that and there may be other use cases for privileged profile.
However, I guess there are some reasons why privileged profile is not provided.
The Kubernetes document describes the following.

https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#node-shell-session
> ・The container runs in the host IPC, Network, and PID namespaces, although the pod isn't privileged, so reading some process information may fail, and chroot /host may fail.
・If you need a privileged pod, create it manually.

So I think it is one of the workaround to create a privileged pod manually.
The example pod manifest is as below.

`pod-node-debug.yaml`
<details>

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: pod-node-debug
  name: pod-node-debug
spec:
  hostIPC: true
  hostNetwork: true
  hostPID: true
  containers:
  - image: <YOUR IMAGE>
    name: debugger
    securityContext:
        privileged: true
    volumeMounts:
    - mountPath: /host
      name: host-root
  volumes:
  - hostPath:
      path: /
      type: ""
    name: host-root
  nodeSelector:
    <NODE LABEL>
```

</details>

**Comment by lion7** [comment]
Thanks a lot for the detailed reply. Talos by default has the [SeccompDefault feature](https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads) enabled, so I guess I should explicitly set the `securityContext` to `Unconfined` to disable it.

I'll try your `sysadmin.yaml` without the `privileged` flag and see what I still can do / cannot do. Manually creating a privileged pod works just fine, it's just that the `kubectl debug` command is more convenient. But I guess there are valid reasons to not support that use-case (perhaps harder to test or prone to human error if it's too convenient).

**Comment by verb** [comment]
The `sysadmin` profile wasn't really intended to mean specifically `CAP_SYS_ADMIN`. It was rather intended to be a profile for system administrators. I don't feel strongly, but I lean towards not creating additional profiles that differ only slightly. If system admins need privileged, then I'd be inclined to add it to the existing profile.

@lion7 thanks for looking into this, I'm happy to have a concrete use case guide this.

**Comment by mochizuki875** [comment]
IMO: Based on the discussion so far, especially [here](https://github.com/kubernetes/kubernetes/pull/119200#issuecomment-1716079280), I think it preferable to add privileged rather than adding specific capability like `CAP_SYS_ADMIN`, `CAP_SYS_PTRACE` etc.

**Comment by mochizuki875** [comment]
@eiffel-fl 
I agree that.
In addition, I think we need to update [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-sysadmin) before change sysadmin profile implementation [same as netadmin](https://github.com/kubernetes/kubernetes/issues/118962#issuecomment-1617538945).

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go
Comment (mochizuki875) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go:233]: I think we should add shareProcessNamespace(pod) in case of podCopy because other profiles like [general](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-general), [baseline](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-baseline) and [restricted](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-restricted) will set shareProcessNamespace by default.

**Review by eiffel-fl — COMMENTED** [review]
Hi.


I normally polished all the code and unit tests, so they should pass.
Regarding the integration one, can you please indicate me what we should do (nothing, changing the cluster permission to run things as admin)?


Best regards.

**Comment by josemotafbn** [comment]
@eiffel-fl - Any way to unblock this testing infra issue ?  
I'm eagerly awaiting this feature :)

**Comment by kfox1111** [comment]
Was needing this today.

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:195]: Why did you change the order?

File: test/cmd/debug.sh
Comment (ardaguclu) [inline test/cmd/debug.sh:290]: reason of this change?

File: test/cmd/debug.sh
Comment (ardaguclu) [inline test/cmd/debug.sh:530]: Could you please check that why the integration test is failing?

**Review by ardaguclu — COMMENTED** [review]
Since the KEP has been merged, we can move forward in this PR.

File: test/cmd/debug.sh
Comment (ardaguclu) [reply test/cmd/debug.sh:529]: I'm not sure there is any strict policy for the integration tests and I'd prefer having these tests unless there is a very strong reasoning. Did you test when you modify this pod yaml as privileged?;

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: target
  name: target
spec:
  securityContext:
    seccompProfile:
      type: Localhost
      localhostProfile: dummy.json
  containers:
  - image: busybox
    name: target
    command: ["/bin/sh", "-c", "sleep 100"]
    securityContext:
        capabilities:
          add:
          - "ALL"
```

Comment (ardaguclu) [reply test/cmd/debug.sh:529]: you can test this also on your local with this command;

```sh
$ make test-integration WHAT=debug_sysadmin
```

Comment (ardaguclu) [reply test/cmd/debug.sh:529]: hmm, it is unfortunate because this integration test would be quite useful. But I think, it is better to skip test, since it is not reasonable to set all cluster as privileged.

Comment (ardaguclu) [reply test/cmd/debug.sh:529]: What about only adding the tests about simply validations, etc. that don't require system level privileges. And of course, with a note why we don't implement all the tests for sysadmin profile?

Comment (ardaguclu) [reply test/cmd/debug.sh:529]: Yes, let's skip the integration tests for sysadmin profile.

Comment (ardaguclu) [reply test/cmd/debug.sh]: After thinking on this, I think we can skip the integration tests @eiffel-fl. We have a good coverage of unit tests which should be sufficient for this case.

**Comment by ardaguclu** [comment]
Thanks!
/triage accepted
/priority backlog
/approve
/lgtm

**Pull request merged**
